### PR TITLE
[NPU] fix some op bugs

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

Bug fixes
### PR changes

OPs
### Describe

fix some op bugs

- square, support int
- update_loss_scaling, fix wrong dtype of output
- concat_grad, support partial empty grad 
- reduce sum, support int
- slice, support nagative starts and ends

### CODE REVIEW

File: paddle/fluid/operators/amp/update_loss_scaling_op_npu.cc [hunk 1]
```diff
@@ -152,7 +152,7 @@ class LazyZerosNPU {
 
         auto place = dev_ctx.GetPlace();
         auto stream = dev_ctx.stream();
-        auto g = out->mutable_data<int>(place);
+        auto g = out->mutable_data<T>(place);
         platform::NPUMemsetAsync(static_cast<void*>(g), 0,
```
Comment (pangyoki) [inline paddle/fluid/operators/amp/update_loss_scaling_op_npu.cc:156]: `sizeof(int)` -> `sizeof(T)`

Comment (zhiqiu) [reply paddle/fluid/operators/amp/update_loss_scaling_op_npu.cc:156]: ok

File: paddle/fluid/operators/concat_op_npu.cc [hunk 2]
```diff
@@ -88,26 +79,39 @@ class ConcatGradNPUKernel : public framework::OpKernel<T> {
...
     axis = ComputeAxis(static_cast<int64_t>(axis),
                        static_cast<int64_t>(ins[0]->dims().size()));
-    // get output tensor that the name is not kEmptyVarName
-    std::vector<framework::Tensor> outputs;
+
     std::vector<int> sizes;
+    int offset = 0;
+    auto stream =
+        ctx.template device_context<paddle::platform::NPUDeviceContext>()
+            .stream();
```
Comment (liym27) [inline paddle/fluid/operators/concat_op_npu.cc:87]: Don't write `stream` by default?

**Review by liym27 — COMMENTED** [review]
Add special tests for these ops to check the bugs are fixed.

File: paddle/fluid/operators/concat_op_npu.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/operators/concat_op_npu.cc:87]: By now, the new feature of default stream is not tested enough. So, I keep the stream now and can remove all streams in the future.

**Review by liym27 — APPROVED** [review]
LGTM

**Pull request merged**